src/charm_config.py

```diff
@@ -0,0 +1,148 @@
+
+    def __init__(self, msg: str):
+        """Initialize a new instance of the CharmConfigInvalidError exception.
+
+        Args:
+            msg (str): Explanation of the error.
+        """
+        self.msg = msg
+
+
+def to_kebab(name: str) -> str:
+    """Converts a snake_case string to kebab-case."""
+    return name.replace("_", "-")
+
+
+class UpfConfig(BaseModel):  # pylint: disable=too-few-public-methods
+    """Represent UPF operator builtin configuration values."""
+
+    model_config = ConfigDict(alias_generator=to_kebab, use_enum_values=True)
+
+    cni_type: CNIType = CNIType.bridge
+    upf_mode: UpfMode = UpfMode.af_packet
+    dnn: StrictStr = Field(default="internet", min_length=1)
+    gnb_subnet: IPvAnyNetwork = IPvAnyNetwork("192.168.251.0/24")
+    access_interface: Optional[StrictStr] = Field(default="")
+    access_interface_mac_address: Optional[StrictStr] = Field(default="")
+    access_ip: str = Field(default="192.168.252.3/24")
+    access_gateway_ip: IPvAnyAddress = IPvAnyAddress("192.168.252.1")
+    access_interface_mtu_size: Optional[int] = Field(
+        default=None, ge=1200, le=65535, validate_default=True
+    )
+    core_interface: Optional[StrictStr] = Field(default="")
+    core_interface_mac_address: Optional[StrictStr] = Field(default="")
+    core_ip: str = Field(default="192.168.250.3/24")
+    core_gateway_ip: IPvAnyAddress = IPvAnyAddress("192.168.250.1")
+    core_interface_mtu_size: Optional[int] = Field(default=None, ge=1200, le=65535)
+    external_upf_hostname: Optional[StrictStr] = Field(default="")
+    vlan_id: Optional[int] = Field(default=None, ge=1, le=4095)
+    enable_hw_checksum: bool = True
+
+    @model_validator(mode="after")
+    def validate_upf_mode_with_mac_addresses(self):
+        """Validate that MAC addresses are defined when in DPDK mode."""
+        if self.upf_mode == "dpdk":
+            invalid_configs = []
+            if not self.access_interface_mac_address:
+                invalid_configs.append("access-interface-mac-address")
+            if not self.core_interface_mac_address:
+                invalid_configs.append("core-interface-mac-address")
+            if invalid_configs:
+                raise ValueError(" ".join(invalid_configs))
+        return self
+
+    @field_validator("access_ip", "core_ip", mode="before")
+    @classmethod
+    def validate_ip_network_address(cls, value: str, info: ValidationInfo) -> str:
+        """Validate that IP network address is valid."""
+        ip_network(value, strict=False)
+        return value
+
+    @field_validator("access_interface_mac_address", "core_interface_mac_address", mode="before")
+    @classmethod
+    def validate_interface_mac_address(cls, value: str, info: ValidationInfo) -> str:
+        """Validate that IP network address is valid."""
+        MacAddress.validate_mac_address(value.encode())
+        return value
+
+
+@dataclasses.dataclass(frozen=True)
+class CharmConfig:
```

Comment:
How about adding a constructor to this class, so that we can return config options separately? If we had that, in the `charm.py`, instead of having a bunch of getters doing `_charm_config.upf_config.some_config_option` we could simply do `_charm_config.some_config_option`. What do you think?